src/charm.py

```diff
@@ -1,104 +1,75 @@
-    """Charm the service."""
+class CinderCSIOperatorCharm(CharmBase):
+    """Deploy and manage the Cinder CSI plugin for K8s on OpenStack."""
 
-    _stored = StoredState()
+    manifests = Path("upstream/manifests")
+    version = Path("upstream/version").read_text()
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
-
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
-
-        Learn more about Pebble layers at https://github.com/canonical/pebble
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
-                }
-            },
-        }
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+        self.cloud_config = CloudConfigRequires(self)
+        self.lk_helpers = LightKubeHelpers(self)
+        self.framework.observe(self.on.install, self._install_or_upgrade)
+        self.framework.observe(self.on.upgrade_charm, self._install_or_upgrade)
+        self.framework.observe(self.on.cloud_config_relation_created, self._install_or_upgrade)
+        self.framework.observe(self.cloud_config.on.ready, self._install_or_upgrade)
+        self.framework.observe(self.on.leader_elected, self._set_version)
+        self.framework.observe(self.on.stop, self._cleanup)
+
+    def _install_or_upgrade(self, event):
+        if not self.cloud_config.relations:
+            self.unit.status = BlockedStatus("Missing cloud-config relation")
+            return
+        if not self.cloud_config.is_ready():
```

Comment:
is there nothing this charm needs from this relation except that it's ready?

Reply:
The relation does provide the MD5 hash of the config contents so that the Cinder charm is notified if the config changes, but other than that, it just needs to know that the `cloud-config` secret has been created in the cluster.